frontend/src/utils/budgetUtil.ts

```diff
@@ -0,0 +1,158 @@
+import { Category, Transaction } from 'src/types/budget';
+
+export interface MonthlyDepositWithdraw {
+  [month: string]: DepositWithdraw;
+}
+
+export interface DepositWithdraw {
+  deposit: number;
+  withdraw: number;
+}
+
+export interface CategoryMap {
+  expense: number;
+  limit: number;
+}
+
+export interface CategoryMaps {
+  [categoryKey: string]: CategoryMap;
+}
+
+export interface CategoryWithdrawMap {
+  [categoryName: string]: number;
+}
+
+export const generateCategoryMaps = (
+  categories: Category[],
+  transactions: Transaction[]
+) => {
+  const result = initializeCategoryMaps(categories);
+  for (const transaction of transactions) {
+    const categoryNames = getCategoryNames(transaction.categoryId, categories);
```

Comment:
I'd create a map of categories before entering the loop and pass the category here directly. This way, you don't have to iterate inside the callee.
```suggestion
    const categoryNames = getCategoryNames(categoriesMap[transaction.categoryId]);
```

Reply:
Changed.

Reply:
Changed